lib/Dialect/HW/Transforms/FlattenIO.cpp

```diff
@@ -0,0 +1,292 @@
+//===- FlattenIO.cpp - HW I/O flattening pass -------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWPasses.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace circt;
+
+static bool isStructType(Type type) {
+  return hw::getCanonicalType(type).isa<hw::StructType>();
+}
+
+// Legal if no in- or output type is a struct.
+static bool isLegalFuncLikeOp(FunctionOpInterface moduleLikeOp) {
+  bool legalResults =
+      llvm::none_of(moduleLikeOp.getResultTypes(), isStructType);
+  bool legalArgs = llvm::none_of(moduleLikeOp.getArgumentTypes(), isStructType);
+
+  return legalResults && legalArgs;
+}
+
+static llvm::SmallVector<Type> getInnerTypes(hw::StructType t) {
+  llvm::SmallVector<Type> inner;
+  t.getInnerTypes(inner);
+  return inner;
+}
+
+namespace {
+
+// Replaces an output op with a new output with flattened (exploded) structs.
+struct OutputOpConversion : public OpConversionPattern<hw::OutputOp> {
+  OutputOpConversion(TypeConverter &typeConverter, MLIRContext *context,
+                     DenseSet<Operation *> *outputOpVisited)
```

Comment:
Why is the set passed as a pointer (instead of a reference)?

Reply:
Since `outputOpVisited` is a mutable value (... code smell, i know) it must be a pointer (referenced cannot be mutable).